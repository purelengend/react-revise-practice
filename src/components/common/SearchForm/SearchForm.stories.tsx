import type { Meta, StoryObj } from "@storybook/react";

// Components
import SearchForm from ".";

const meta: Meta<typeof SearchForm> = {
  title: "Components/common/SearchForm",
  component: SearchForm,
  argTypes: {
    onSubmit: {
      description: "The submit callback of the form",
    },
  },
  args: {
    onSubmit: () => {},
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof SearchForm>;

export const Primary: Story = {};
