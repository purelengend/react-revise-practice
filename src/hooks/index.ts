export * from "./useStudent";
export * from "./useImage";
export * from "./useDebounce";
export * from "./useQueryParams";
export * from "./useStudentModal";
