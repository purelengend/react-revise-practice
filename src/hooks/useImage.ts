import { useMutation } from "@tanstack/react-query";

// Services
import { uploadImageAndGetUrl } from "@/services";

// Constants
import { DEFAULT_AVATAR_URL } from "@/constants";
import { useCallback } from "react";

export const useImage = () => {
  const {
    data: image,
    mutate,
    isPending: isUploadingImage,
  } = useMutation({
    mutationFn: (data: FormData) => uploadImageAndGetUrl(data),
  });

  const uploadImage = useCallback(
    (image: File) => {
      const imageFormData = new FormData();

      imageFormData.append("image", image);

      mutate(imageFormData);
    },
    [mutate],
  );

  return {
    uploadImage,
    isUploadingImage,
    imageUrl: image ? image.data.data.url : DEFAULT_AVATAR_URL,
  };
};
