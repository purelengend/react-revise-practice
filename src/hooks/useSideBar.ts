import { useDisclosure } from "@chakra-ui/react";

export const useSidebar = () => {
  const { isOpen, onOpen, onClose, onToggle } = useDisclosure({
    defaultIsOpen: true,
  });

  return {
    sidebarState: isOpen,
    openSidebar: onOpen,
    closeSidebar: onClose,
    toggleSidebar: onToggle,
  };
};
