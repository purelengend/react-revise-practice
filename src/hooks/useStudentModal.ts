import { useState } from "react";

// Constants
import { DEFAULT_STUDENT_DATA } from "@/constants";

// Types
import { Student } from "@/types";

export const useStudentModal = () => {
  const [student, setStudent] = useState<Student>(DEFAULT_STUDENT_DATA);

  const [studentId, setStudentId] = useState<string>("");

  return {
    student,
    setStudent,
    studentId,
    setStudentId,
  };
};
