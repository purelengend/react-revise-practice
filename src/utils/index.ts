export * from "./test-utils";
export * from "./config";
export * from "./customToast";
