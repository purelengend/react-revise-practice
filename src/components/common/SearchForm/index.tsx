import {
  IconButton,
  Input,
  InputGroup,
  InputRightElement,
} from "@chakra-ui/react";
import { Controller, useForm } from "react-hook-form";
import { useCallback, useEffect, useState } from "react";

// Components
import { SearchIcon } from "../Icons";

// Constants
import { SEARCH_INPUT_PLACEHOLDER } from "@/constants";

// Hooks
import { useDebounce } from "@/hooks";

export type SearchFormProps = {
  onSubmit: (args: SearchFormInput) => void;
  searchParam: string;
};

export type SearchFormInput = {
  searchValue: string;
};
const SearchForm = ({ onSubmit, searchParam }: SearchFormProps) => {
  const { control, handleSubmit } = useForm<SearchFormInput>({
    defaultValues: {
      searchValue: "",
    },
  });

  const [searchText, setSearchText] = useState("");

  const debouncedText = useDebounce(searchText.replace(/\s+/g, " ").trim()); // remove extra spaces of the search text

  const handleOnchange = useCallback(
    (
      event: React.ChangeEvent<HTMLInputElement>,
      onChange: (...event: unknown[]) => void,
    ) => {
      onChange(event);
      setSearchText(event.target.value);
    },
    [],
  );

  const checkKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>) => {
      if (event.key === "Enter") event.preventDefault();
    },
    [],
  );

  useEffect(() => {
    if (searchParam !== debouncedText) handleSubmit(onSubmit)();
  }, [debouncedText, handleSubmit, onSubmit, searchParam]);

  return (
    <InputGroup
      as="form"
      onSubmit={handleSubmit(onSubmit)}
      onKeyDown={checkKeyDown}
    >
      <Controller
        name="searchValue"
        control={control}
        render={({ field: { onChange, ...rest } }) => (
          <Input
            {...rest}
            h={9.25}
            placeholder={SEARCH_INPUT_PLACEHOLDER}
            color="black"
            onChange={(event) => handleOnchange(event, onChange)}
          />
        )}
      />
      <InputRightElement>
        <IconButton
          as="button"
          type="submit"
          aria-label="Search student"
          _hover={{
            bg: "none",
          }}
          icon={<SearchIcon />}
        />
      </InputRightElement>
    </InputGroup>
  );
};

export default SearchForm;
