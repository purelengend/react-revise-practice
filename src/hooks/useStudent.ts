import {
  keepPreviousData,
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query";
import { useContext } from "react";
import { useToast } from "@chakra-ui/react";

// Services
import {
  createOrUpdateStudent,
  deleteStudentById,
  getAllStudents,
  getStudentById,
} from "@/services";

// Context
import { UrlContext } from "@/context";

// Types
import { Student } from "@/types";

// Constants
import { QUERY_KEY, TOAST_MSG, TOAST_STATUS } from "@/constants";

//Utils
import { customToast } from "@/utils";

export const useStudent = (id: string = "") => {
  const { path } = useContext(UrlContext);

  const queryClient = useQueryClient();

  const toast = useToast();

  const {
    data: students,
    refetch: refetchStudents,
    isFetching: isFetchingStudentData,
  } = useQuery({
    queryKey: [QUERY_KEY.STUDENT, path],
    queryFn: () => {
      return getAllStudents(path);
    },
    placeholderData: keepPreviousData,
  });

  const {
    mutate: mutateStudent,
    isPending: isMutatingStudent,
    isSuccess: isMutateStudentSuccess,
  } = useMutation({
    mutationKey: [QUERY_KEY.STUDENT, "mutation"],
    mutationFn: (data: Student) => createOrUpdateStudent(data),
    onSuccess: (_, { id }) => {
      if (id) {
        toast(
          customToast(
            TOAST_MSG.ADD.SUCCESS.title,
            TOAST_MSG.ADD.SUCCESS.description,
            TOAST_STATUS.SUCCESS,
          ),
        );
        return;
      }

      toast(
        customToast(
          TOAST_MSG.EDIT.SUCCESS.title,
          TOAST_MSG.EDIT.SUCCESS.description,
          TOAST_STATUS.SUCCESS,
        ),
      );

      queryClient.invalidateQueries({ queryKey: [QUERY_KEY.STUDENT, path] });
    },
    onError: (_, { id }) => {
      if (id) {
        toast(
          customToast(
            TOAST_MSG.ADD.ERROR.title,
            TOAST_MSG.ADD.ERROR.description,
            TOAST_STATUS.ERROR,
          ),
        );
        return;
      }
      toast(
        customToast(
          TOAST_MSG.EDIT.ERROR.title,
          TOAST_MSG.EDIT.ERROR.description,
          TOAST_STATUS.ERROR,
        ),
      );
    },
  });

  const { data: studentByIdData } = useQuery({
    queryKey: [QUERY_KEY.STUDENT, id],
    queryFn: () => getStudentById(id),
  });

  const {
    mutate: deleteStudent,
    isPending: isDeletingStudent,
    isSuccess: isDeleteStudentSuccess,
  } = useMutation({
    mutationKey: [QUERY_KEY.STUDENT, "delete"],
    mutationFn: (id: string) => deleteStudentById(id),
    onSuccess: () => {
      toast(
        customToast(
          TOAST_MSG.DELETE.SUCCESS.title,
          TOAST_MSG.DELETE.SUCCESS.description,
          TOAST_STATUS.SUCCESS,
        ),
      );

      queryClient.invalidateQueries({ queryKey: [QUERY_KEY.STUDENT, path] });
    },
    onError: () => {
      toast(
        customToast(
          TOAST_MSG.DELETE.ERROR.title,
          TOAST_MSG.DELETE.ERROR.description,
          TOAST_STATUS.ERROR,
        ),
      );
    },
  });
  return {
    students,
    refetchStudents,
    isFetchingStudentData,
    mutateStudent,
    isMutatingStudent,
    isMutateStudentSuccess,
    studentByIdData,
    deleteStudent,
    isDeletingStudent,
    isDeleteStudentSuccess,
  };
};
