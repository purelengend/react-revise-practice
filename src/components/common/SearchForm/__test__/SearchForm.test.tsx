import { render, renderHook, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// Components
import SearchForm from "..";

// Hooks
import { useQueryParams } from "@/hooks";

// Utils
import { AllTheProviders } from "@/utils";

describe("SearchForm test cases", () => {
  const mockOnSubmit = jest.fn();

  const setup = () => {
    userEvent.setup();
    return render(<SearchForm searchParam="mock" onSubmit={mockOnSubmit} />);
  };
  it("should render correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("should invoke onChange function when typing", async () => {
    setup();

    const searchInput = screen.getByRole("textbox");

    await userEvent.type(searchInput, "mock search");

    expect(searchInput).toHaveValue("mock search");
  });

  it("should prevent the enter key from being pressed", async () => {
    setup();

    const { result } = renderHook(() => useQueryParams(), {
      wrapper: AllTheProviders,
    });

    const searchBtn = screen.getByRole("button", {
      name: /search student/i,
    });

    await userEvent.type(searchBtn, "{enter}");

    expect(result.current.name).toBe("");
  });
});
