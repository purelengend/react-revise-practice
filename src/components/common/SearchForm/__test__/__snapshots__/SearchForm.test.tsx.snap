// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm test cases should render correctly 1`] = `
<div>
  <form
    class="chakra-input__group css-1utxn56"
    data-group="true"
    id="search-form"
  >
    <div
      class="css-79elbk"
    >
      <input
        class="chakra-input css-sh2byc"
        name="searchValue"
        placeholder="Search..."
        value=""
      />
    </div>
    <div
      class="chakra-input__right-element css-dp9r5f"
    >
      <button
        aria-label="Search student"
        class="chakra-button css-15ql2fp"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-1vc1jog"
          focusable="false"
          viewBox="0 0 14 14"
        >
          <g
            clip-path="url(#clip0_63_207)"
            id="search 1"
          >
            <path
              d="M13.9043 13.1687L10.377 9.64141C10.3141 9.57852 10.232 9.5457 10.1445 9.5457H9.86289C10.8008 8.53125 11.375 7.17773 11.375 5.6875C11.375 2.5457 8.8293 0 5.6875 0C2.5457 0 0 2.5457 0 5.6875C0 8.8293 2.5457 11.375 5.6875 11.375C7.17773 11.375 8.53125 10.8008 9.5457 9.86562V10.1445C9.5457 10.232 9.58125 10.3141 9.64141 10.377L13.1688 13.9043C13.2973 14.0328 13.5051 14.0328 13.6336 13.9043L13.9043 13.6336C14.0328 13.5051 14.0328 13.2973 13.9043 13.1687ZM5.6875 10.5C3.02695 10.5 0.875 8.34805 0.875 5.6875C0.875 3.02695 3.02695 0.875 5.6875 0.875C8.34805 0.875 10.5 3.02695 10.5 5.6875C10.5 8.34805 8.34805 10.5 5.6875 10.5Z"
              fill="currentColor"
              id="Vector"
            />
          </g>
          <defs>
            <clippath
              id="clip0_63_207"
            >
              <rect
                fill="white"
                height="14"
                width="14"
              />
            </clippath>
          </defs>
        </svg>
      </button>
    </div>
  </form>
</div>
`;
